### ISSUE TITLE
GameGuru MAX - Request - Project Folders

### ISSUE THREAD

**Issue by ArgentArts** [TheGameCreators/GameGuruRepo#2349]
I'd like to see the idea of projects taken to the next level. I do a lot of testing with MAX and probably will, even when I get serious with developing a game in it. As such, I end up with game projects and game levels that I've used for testing purposes that I need to easily remove, all filled with temporary models that I also need to easily remove. But it's not so easy. There's really no real way to delete anything in MAX. In fact, deleting a level from your Storyboard ONLY removes it from your Storyboard (the actual level still remains saved in your mapbank). Ultimately, what you end up with is a mess of maps and models and materials all jumbled together from various projects. It doesn't matter if you have one or three or a dozen projects that you're working on, all the levels, all the models, they are all saved in the same place. And if you decide you no longer need one of the projects and want to totally remove all of it's levels, models, materials, sounds, etc., you have to go through every single folder and hunt for it's particular unique items to remove or delete them. This is a PITA, to say the least.

I would like to propose that GameGuru MAX creates a totally separate project folder for each new project. This project folder can be in the documents folder or in the user folder or wherever. It can be titled after the game/project name to make it easy to find. Any unique elements (maps, user imported models, materials, sounds, etc.) would get stored in this project folder within it's proper directory. This way, if the end-user ever wanted to delete a particular project, all they would have to do is delete that project folder and it's done. The project folder would keep everything separate from the main MAX folders, etc., so there would be no danger of removing anything relating to MAX supplied content. The project folder would keep everything separate from other project folders so there would be no worries of removing anything related to other projects.

Having something like this would allow the end-user to organize their projects for easy backups and allow them to easily remove unwanted projects, too. Concerning backups - this is a pretty big one. Having a project folder like this would allow the end-user to create iterative backups, allowing them to return to previous "builds" should they need or want to. 

Right now, as an alpha/beta tester, I bring in a lot of models to test this or that. I don't intend to keep or use these models for anything other than testing MAX. After conducting this test or that, I go into the entitybank and I have to hunt for all of the models to find which ones to delete. Then I have to go through the mapbank and project bank and get rid of the ones I no longer need there. All because there is no way to manage them from within MAX and all because there is no folder created for them to keep them all separate from each other when initially created.

So, please, create a proper means of keeping projects separate in MAX for those of us that may be working on more than one project at a time. :)

**Comment by ArgentArts** [comment]
Yep. As I see it, a project folder would contain copies of anything that's not unique to MAX or to another project AND anything that is unique to itself.

**Comment by Necrym59** [comment]
Yes i agree I have made this request earlier, even suggesting keeping the same structure but just under project directory names. Its such a no brainer that it should be done. It would stop all the cross file issues and make a cleaner control of projects.
https://github.com/TheGameCreators/GameGuruRepo/issues/1841

**Comment by MonkeyFrogStudio** [comment]
@LeeBamberTGC Sorry to tag you, but this was assigned to you. ;) Can this not be forgotten? The GameGuru MAX folder structure is a mess. I have so many tests going on for testing various bugs for MAX, creating a demo, and working on my own project(s) ... it's just one big jumble in my user folder. When I go to delete an old test, I have to hut down everything in this folder and that (entitybank, levelbank, mapbank, projectbank) and hope I've gotten it all (and not accidently removed something I wasn't supposed to. 

MAX acts like it wants to have projects. It makes us name stuff before we move to the next stage (name the terrain it generates, name the level before going to the editor, etc.). Why not just go ahead and name the  project before we even begin? Let that create a main folder and the folder structure there to put all unique files in for that project. And if the end-user ever wants to delete that project or back that project up, they just have to delete that ONE folder. If we ever have to send you a project to examine, we know it would be that ONE folder (or at least in that one folder). 

I think this would simplify everyone's lives.

**Comment by MonkeyFrogStudio** [comment]
@LeeBamberTGC @davetgc There are several of us that would really like to see this in MAX sooner rather than later. It's a bit of a PITA to have to go into your Files area and go to both mapbank and projectbank to delete everything. I am often testing things and, when done, I'd like to easily just delete my tests in one stroke. If we had actual project folders, then I could, for example, create a Test project and then delete when my testing is done.

**Comment by nomad5oul** [comment]
I would also really like to see this. It makes a lot of sense and will be much better way of working in Max.
Hope this one gets a green light.

**Comment by synchromesh62** [comment]
Must admit It would be a lot more organised and cleaner if each project folder created its own file structure. :)

**Comment by Granada1** [comment]
Have to agree , greatly need please 🙏

**Comment by LeeBamberTGC** [comment]
@MonkeyFrogStudio I agree, I think this will be needed for anyone even half-serious about their projects, and the ability to delete and archive projects from the HUB to manage projects too.

As a first step, how about we add a THIRD folder location when in a game project. We have the MAX root (read only), and then Writeables folder (GameGuruApp) and your game project would be the third (i.e. "Files\projectbank\Trapped\Files\entitybank").  In this way any media files can be stored local to the project but discoverable with all the other sources.

We can then change the edit icon to a [...] more button that has a drop-down to Edit Project, Delete Project (with confirm) and Archive Project (which confirms then places it in a backup folder for safekeeping).

Add a new filter called "Show Archived Projects" that can then be restored by clicking it to return it to the HUB:

![image](https://user-images.githubusercontent.com/35840583/213867602-c2309ad5-e9db-4f2d-b713-2d817157dde5.png)

Have I missed anything out?

**Comment by MonkeyFrogStudio** [comment]
Hi @LeeBamberTGC 
That all sounds really good, to me at least. 
One thing - I think many of us are thinking that the newly created project folder would not just have an entitybank folder in it, but all the needed folders (scriptbank, audiobank, mapbank, etc.). I am sure you intended this, but wanted to make sure this was the case. That way, if the end-user wanted, everything for a single game project could be self-contained in the one project folder.

**Comment by LeeBamberTGC** [comment]
@MonkeyFrogStudio Just so, it would have all contributing custom media in the Files folder.  We may need to add an icon to the browsers so the user is always aware which ones are stock library, which ones are common writables and which are project specific.  I can see a scenario where a user adds a game project import, creates a new project and wanders why their object is no longer in the library :)

**Comment by MonkeyFrogStudio** [comment]
@LeeBamberTGC Could there be some kind of "referencing" (sorry if I am not using the right term here ... artist, not programmer) so that if the end-user uses an asset that is, say, in the default MAX library, instead of it being copied into the Project Folder, creating a duplicate and wasting hard drive space, some sort of reference (like a table) would be used to track assets used outside the project library? This way, even if you archive the project folder and restore it later, it would still keep track of those assets used, too.

**Comment by LeeBamberTGC** [comment]
@MonkeyFrogStudio Yes, objects taken from the stock or writables would keep those references (no file copying needed). The third files folder in the project would be the only transient folder.  A typical level may well contain objects from all three locations which is fine as the level files will now be stored inside the project folder (perhaps where they should have always been) :)

**Comment by MonkeyFrogStudio** [comment]
@LeeBamberTGC Sounds great to me!

**Comment by Granada1** [comment]
Just what the doctor ordered, sounds great 👍

**Comment by synchromesh62** [comment]
Excellent :)

**Comment by nomad5oul** [comment]
Awesome news !!

**Comment by LeeBamberTGC** [comment]
ALL: new EXP build uploaded with support for project media folders.  Just open your project once and you will find a new Files folder with appropriate subfolders. Only place your media in the "Your Project Files" subfolders manually for now. Eventually, we will have controls in the UI (i.e importer) to direct your custom content there.

Also, the MAPBANK code is referenced in around 100 different places so this build does not mess with moving the FPM location for now. Let's see how much this has messed things up before taking the next step.

I have renamed this task to switch the enhancement to focus on MAPBANK to complete this. Let me know if anything bad happens with this one. Remember to back up regularly when dealing with Early Access Software!

**Comment by MonkeyFrogStudio** [comment]
@LeeBamberTGC Now MAX is crashing whenever it tries to load my game project, making MAX unusable on my end.

**Comment by LeeBamberTGC** [comment]
@MonkeyFrogStudio Darn that was fast :) Beat me to to it by 10 seconds, fix uploaded. Too many entries in the change log, the announcer was not designed to report so much :)

**Comment by Tum1370** [comment]
Latest EXP build crashes as soon as opening a project now.

**Comment by MonkeyFrogStudio** [comment]
@LeeBamberTGC Still crashing on both old and new projects.

**Comment by LeeBamberTGC** [comment]
@MonkeyFrogStudio Send me your changelog.txt file :)

**Comment by MonkeyFrogStudio** [comment]
@LeeBamberTGC 

[changelog.txt](https://github.com/TheGameCreators/GameGuruRepo/files/10609121/changelog.txt)

**Comment by MonkeyFrogStudio** [comment]
@LeeBamberTGC I managed to get it to work. I had to revert to a previous version of MAX, run it, then go back to experimental. Now I can load my project.

**Comment by MonkeyFrogStudio** [comment]
@LeeBamberTGC Are the "Your Project Files" folder required? Or are these something we can remove?

Also, instead of this being saved in the projectbank folder, can we have it so we can designate a place to create the project folder when we create a new project?

**Comment by LeeBamberTGC** [comment]
@MonkeyFrogStudio Looks fine. Maybe send me one of your new projects created in this version, thanks!  I have attached one of mine so you can test.
[New.zip](https://github.com/TheGameCreators/GameGuruRepo/files/10609136/New.zip)

**Comment by MonkeyFrogStudio** [comment]
@LeeBamberTGC Read up a few posts. :) I was able to sort it out (and asked a few questions). :)

**Comment by LeeBamberTGC** [comment]
@MonkeyFrogStudio The "Your Project Files" are automatically created for any project, your media needs to go in there so they do not conflict with any other root or DLC media folders that may arrive in the future.

**Comment by LeeBamberTGC** [comment]
@MonkeyFrogStudio Projects need to go in the projectbank folder, that is the purpose of the folder :)

**Comment by MonkeyFrogStudio** [comment]
> @MonkeyFrogStudio Projects need to go in the projectbank folder, that is the purpose of the folder :)

@LeeBamberTGC

I get that. However, no other game engine requires such a thing (at least not Unity, Unreal, Flax, S2, etc.). They all allow the flexibility to place a project wherever you like on your hard drive(s). When testing features and new ideas, I often will create a "project" in a folder I've named "temp". I know that anything in my "temp" folder can be safely deleted. It would be great if I could designate where to place my own projects instead of being forced to use the projectbank which is forced to be in My Documents. Many of us have several hard drives and it would be great if we could designate where we want our MAX projects to be stored and read from.

**Comment by Tum1370** [comment]
@LeeBamberTGC 
The latest build creates a empty "Files" folder in the Projectbank folder as well. Seem a bug as not needed ?
Also, Inside the Project and the Files folders, why also add the "Your Project Files" inside each media folder ? 
Seems to be just another wasted layer of folder ?

**Comment by LeeBamberTGC** [comment]
@MonkeyFrogStudio You can change the location of your files using the Advanced setting here: 

![image](https://user-images.githubusercontent.com/35840583/216820249-bdd0b911-3557-4c6c-86ef-9323ac332b05.png)

**Comment by MonkeyFrogStudio** [comment]
@LeeBamberTGC Thanks! Not sure how I overlooked that! :)

This is great. However, with most game creation applications, instead of it being a setting that effects every project, this should actually be a "per project" setting, imo. So, instead of having to go to Settings and choosing a directory for where to save all my projects, it should work like this (I think):

- Create a new project and give it a project name
- MAX then prompts me for a place to save the project
- MAX then creates the required folder structure under the given project name at the location provided in the above step

Therefore, if I create a new project called Test, MAX would then ask me where to save Test. I could then navigate to D:/My Cool Test Folder (or wherever) and hit Okay. After that, in D:/My Cool Test Folder there would be created Test with all the the required subfolders. If, later, I wanted to create another project ... perhaps something serious that's not a test ... I could create the project and save it to a completely different place, such as D:/Work/GameProjects/MyCoolGame. Whenever I would open Test, any changes would go to D:/My Cool Test Folder/Test. Whenever I opened MyCoolGame, it would automatically know to save to D:/Work/GameProjects/MyCoolGame.

**Comment by MonkeyFrogStudio** [comment]
@LeeBamberTGC BTW, I don't want you to think I am ungrateful. I think it's amazing that MAX is getting project folders! Thanks for taking the time to do this. It's an amazing step forward.

**Comment by LeeBamberTGC** [comment]
@MonkeyFrogStudio Sounds like what we need is a sort of project-specific writable area. We would need to retain some common writables location for settings and suchlike, but we could create a project-specific writables area for each project created.

So by default it works as now so most users can go to one location to find all their projects (the current writables), but when an "Advanced mode feature" is ticked to require a separate project folder per game project, an extra dialog is presented when you create a new project to ask where you would like it stored. Sound about right?

**Comment by MonkeyFrogStudio** [comment]
@LeeBamberTGC If it allows the end-user to select any drive, anywhere to save their project (and MAX will create the required folder structure, etc.), then sure. 

Having said that, I personally don't see why this needs to be an advanced feature. Just about every program on the planet works by allowing the end-user to select exactly where to save their own files without having to access an Advanced Mode to enable such a feature. This means it's standard practice (and, therefore, common to every end-user, no matter how skilled) to simply select where to save your files. 

Now, a game creation system, like MAX, is more complex (i.e. more moving parts - models, materials, shaders, scripts, etc.), so instead of telling MAX where to save a single file, per se, the end-user would be telling MAX where to set up the project (and MAX would then create the file structure it requires there).

Again, just about any game engine I can think of functions this way. I still have folders from tests I'd done in other engines in my custom Temp director - Flax, S2, O3DE, Unreal, etc. But I create all of my non-test stuff in my Work/Game Development folder on my D drive.

**Comment by MonkeyFrogStudio** [comment]
For ease of reference, here are a few links to topics related to this thread:

https://github.com/TheGameCreators/GameGuruRepo/issues/2087

https://github.com/TheGameCreators/GameGuruRepo/issues/2214

**Comment by OldFlak** [comment]
Personally, I don't see why we need to complicate the whole folder setup - the Custom Folder Location is already a USER folder.
As long as everything is in the correct folder ie, scripts etc, then we should be able to have sub-folders to our own discretion.

All these **User** folders and **Your Project Files** folder requirements don't make sense to me for the above stated reason.

The Project folders should simply follow the same rule. And should include any folders that can use custom media.

![image](https://user-images.githubusercontent.com/12246318/218347087-ca2a8292-6fc0-4a13-8e32-d27378940d62.png)

**Comment by LeeBamberTGC** [comment]
@OldFlak Okay removed the need for the mad "Your Project Files" folder, and added a new selection to the Library system called "Project Media" which simply lists any media you have in your local project folder structure as suggested above:

![image](https://user-images.githubusercontent.com/35840583/219092253-a1f425dc-b0d3-4a2c-a775-a80e03ce1360.png)
